AWS: Support . in tectonic_domain; make DNS consistent

joshrosso opened the pull request:
This PR attempts to achieve the following in the terraform installer for AWS.

**a.** Allow `.`s in `tectonic_domain`.
**b.** Ensure consistency of DNS records created in Route 53.

## Details

**a**: 
- `.`s are allowed in route53 resource record sets. 
- However, since the custom domain specified (variable: tectonic_domain) [is used to name the ELBs provisioned by tectonic](https://github.com/coreos/tectonic-installer/blob/12df2696f4968e03246b3574c36cbed33ca198f7/modules/aws/master-asg/elb.tf#L92), terraform will fail due to AWS only allowing `[a-zA-Z0-9-]+` in ELB names. The following error is the result.
  ```
  2 error(s) occurred:
  * module.masters.aws_elb.api-external: only alphanumeric characters and hyphens allowed in "name": "josh.calico-ext"
  * module.masters.aws_elb.console: only alphanumeric characters and hyphens allowed in "name": "josh.calico-con"
  ```
- This PR uses the cluster-name **exclusively** when naming the ELBs.
  - I considered sanitizing the `custom_domain_name` (e.g. `.`->`-`), but after some thought, decided consistency with the cluster name over DNS name is likely preferred. We can change this if we disagree.

**b**:
- The formatting of DNS records created by terraform is somewhat inconsistent and hard to correlate. 
- The SRV records specifically are generic and can cause collisions if 2 clusters wish to use the same hosted zone. 
- This commit ensures all records follow the syntax.
  ```
  ${IDENTIFIER}.${TECTONIC_DOMAIN}.${BASE_DOMAIN}
  ```

## Example

The following variables

`tectonic_domain` = `tectonic.team1`
`base_domain` = `coreos.com`

Now result in the following records created.

```
console:
tectonic.team1.coreos.com

api server:
api.tectonic.team1.coreos.com

etcd:
etcd-0.tectonic.team1.coreos.com
etcd-1.tectonic.team1.coreos.com
etcd-2.tectonic.team1.coreos.com

etcd-srv:
_etcd-client-ssl._tcp.tectonic.team1.coreos.com
_etcd-server-ssl._tcp.tectonic.team1.coreos.com
```

joshrosso commented on platforms/aws/tectonic.tf:
```diff
@@ -30,13 +30,13 @@ module "bootkube" {
-    "${var.tectonic_cluster_name}-etcd-4.${var.tectonic_base_domain}",
-    "${var.tectonic_cluster_name}-etcd-5.${var.tectonic_base_domain}",
-    "${var.tectonic_cluster_name}-etcd-6.${var.tectonic_base_domain}",
+    "etcd-0.${var.tectonic_dns_name == "" ? var.tectonic_cluster_name : var.tectonic_dns_name}.${var.tectonic_base_domain}",
```
Is there a more elegant way to do this rather than a bunch of ifs?

alekssaul replied:
doesn't look like `etcd-0.${var.tectonic_dns_name == ""` will ever be true. If `$var.tectonic_dns_name` is null, `etcd-0.${var.tectonic_dns_name}` will be "etcd-0." no?

s-urbaniak replied:
curently no :-( I think it is time for custom datasources.

joshrosso commented:
In the smoke tests, it's failing on baremetal.

Based on my observations, the matchbox server is never reached.

```
++ curl --silent --fail -k http://matchbox.example.com:8080
+ echo 'Waiting for matchbox...'
Waiting for matchbox...
+ sleep 5
++ curl --silent --fail -k http://matchbox.example.com:8080
+ echo 'Waiting for matchbox...'
Waiting for matchbox...
+ sleep 5
Sending interrupt signal to process
/jenkins/workspace/tectonic-installer_PR-1295-QAGWK67ZOXX7HRMKSAAXMBWRK6DK6Y5BC3MAQTPFUFNT4A2KR4BA/tests/smoke/bare-metal/smoke.sh: line 42: 26606 Terminated              sleep 5
+ kill_terraform_and_cleanup
+ echo 'Killing Terraform'
Killing Terraform
+ kill
kill: usage: kill [-s sigspec | -n signum | -sigspec] pid | jobspec ... or kill -l [sigspec]
+ true
+ echo 'WARNING: Cleanup is temporarily disabled on failure for debugging purposes. Next job will clean at startup.'
WARNING: Cleanup is temporarily disabled on failure for debugging purposes. Next job will clean at startup.
script returned exit code 143
```

joshrosso commented:
I will knock out **a** and add it to this PR today.

**b** is likely too much for me to commit to at the moment, especially when it comes to testing and such. If we feel ok with opening up an issue around DNS consistency and tackling it when we can, that'd be my preference.

Thanks @s-urbaniak

s-urbaniak commented:
@joshrosso the plan SGTM!

robszumski commented:
@joshrosso just to be clear, are you going to do A for all platforms? I'd really really like to start enforcing changes like this to be made across all platforms or not at all. It is going to get crazy if we aren't strict about this. Moving B to another issue sounds good to me.

sym3tri commented on modules/aws/master-asg/elb.tf:
```diff
@@ -32,7 +32,7 @@ resource "aws_elb" "api-internal" {
 resource "aws_route53_record" "api-internal" {
   zone_id = "${var.internal_zone_id}"
-  name    = "${var.custom_dns_name == "" ? var.cluster_name : var.custom_dns_name}-api.${var.base_domain}"
+  name    = "api.${var.custom_dns_name == "" ? var.cluster_name : var.custom_dns_name}.${var.base_domain}"
```
We may run into similar issues here with ELB maxchar naming limits that we had with long cluster names.

@s-urbaniak @squat thoughts on how to address?

joshrosso commented:
I've rebased this PR. Sorry to see it slipped over time.

I've removed all the SRV etcd mentions from the code as it appears we're no longer using SRV records for initial etcd bootstrapping.

Based on feedback and current project state, appears the following is still needed. Some questions below.

1. Review how this affects the GUI and make sure it still works as expected.
    - Who is responsible / best for this?
2. Document the conventions used for resource naming and DNS name construction.
    - I can do this, where does it belong?
3. Update our other "stable" platforms (currently bare-metal only) to follow the same conventions.
    - I can do this in my free time.

cpanato commented:
closing due to inactivity. thanks for your PR. if you think this is still useful, please re submit.
